_posts/2021-02-20-kotlin-use-site-targets.md

```diff
@@ -0,0 +1,83 @@
+---
+layout: post
+title:  "Unwrapping Use-site Targets in Kotlin"
+date:   2021-03-02 10:00:00 +0200
+comments: true
+img:
+  href: 2021-03-02-dogs-of-berlin.png
+  copyright: Dogs of Berlin by Christian Alvart
+  alt: Dogs of Berlin
+---
+I like Kotlin for its expressiveness. Its interoperability with Java
+allows to easily use the existing Java ecosystem while enjoying a modern
+language. When it comes to annotations Kotlin handles them almost the same
+as Java. In Kotlin, a single syntax construction might be compiled into
+several JVM constructions. A Kotlin property would produce a JVM field,
+a getter, and optionally a setter. This is where annotation use-site targets
+are useful.
+
+{% include picture.html %}
+
+Take a look at the following code:
+
+```kotlin
+data class User(
+    @Unique
+    val name: String,
+    @Size(min = 18, max = 100)
+    val age: Int,
+)
+```
+
+Once it is compiled, the corresponding class would have a constructor,
+fields, getters, equals, hashcode, and some other language elements.
+If you need that in the compiled code annotations were generated specifically
+for getters or fields you can specify so-called use-site targets:
+
+```kotlin
+data class User(
+    @field:Unique
+    val name: String,
+    @get:Size(min = 18, max = 100)
+    val age: Int,
+)
+```
+In the compiled code the `name` field will be annotated with `@Unique`
+and a getter of the `age` property will have the `@Size` annotation.
+
+This all works great unless you forget to put the corresponding target in front
```

Comment:
"All this works great, as long as you don't forget to put the corresponding target in front of an annotation.